compiler/rustc_middle/src/lint.rs

```diff
@@ -257,7 +257,11 @@ pub fn struct_lint_level<'s, 'd>(
     level: Level,
     src: LintLevelSource,
     span: Option<MultiSpan>,
-    decorate: impl for<'a> FnOnce(LintDiagnosticBuilder<'a, ()>) + 'd,
+    msg: impl Into<DiagnosticMessage>,
+    decorate: impl 'd
+    + for<'a, 'b> FnOnce(
+        &'b mut DiagnosticBuilder<'a, ()>,
+    ) -> &'b mut DiagnosticBuilder<'a, ()>,
```

Comment:
Why do we need `'s` and `'d`?  Can we get rid of them and get a nicer impl-trait?

Reply:
Also: why do we need to return the `&mut DiagnosticBuilder`? Can't `decorate` just return unit?

Reply:
> Why do we need `'s` and `'d`? Can we get rid of them and get a nicer impl-trait?

I don't immediately see why we'd need them, probably an old artifact of sorts. I'll try removing it.

Reply:
> Also: why do we need to return the &mut DiagnosticBuilder? Can't decorate just return unit?

It absolutely can. I chose this signature so that the most common (from what I've gathered while refactoring) case of "just 1-2 things to add" is nicer, since all `DiagnosticBuilder` methods return `&mut DiagnosticBuilder`, chains just work with this API.
```rust
cx.struct_span_lint(
    LINT, 
    span,
    fluent::lint::blah,
    |lint| lint.set_arg("ty_param", ty_param).span_note(args_span, fluent::lint::note),
);

// vs

cx.struct_span_lint(
    LINT, 
    span,
    fluent::lint::blah,
    |lint| { lint.set_arg("ty_param", ty_param).span_note(args_span, fluent::lint::note); },
);
```

Additionally, maybe even more importantly, returning the builder it shows that you are not expected to `.emit()` it.